pkg/virt-controller/watch/migration.go

```diff
@@ -772,13 +806,347 @@ func (c *MigrationController) listMigrationsMatchingVMI(namespace string, name s
+	c.pdbExpectations.CreationObserved(vmiKey)
+
+	if err := c.syncPDB(vmi); err != nil {
+		log.Log.V(2).Reason(err).Errorf("could not sync PDB for vmi %v on added PDB %v", vmi.Name, pdb.Name)
+	}
+}
+
+func (c *MigrationController) deletePodDisruptionBudget(obj interface{}) {
+	pdb, ok := obj.(*v1beta1.PodDisruptionBudget)
+
+	// When a delete is dropped, the relist will notice a pdb in the store not
+	// in the list, leading to the insertion of a tombstone object which contains
+	// the deleted key/value. Note that this value might be stale. If the pdb
+	// changed labels the new vmi will not be woken up till the periodic resync.
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			log.Log.Reason(fmt.Errorf("couldn't get object from tombstone %+v", obj)).Error("Failed to process delete notification")
+			return
+		}
+		pdb, ok = tombstone.Obj.(*v1beta1.PodDisruptionBudget)
+		if !ok {
+			log.Log.Reason(fmt.Errorf("tombstone contained object that is not a pdb %#v", obj)).Error("Failed to process delete notification")
+			return
+		}
+	}
+
+	controllerRef := v1.GetControllerOf(pdb)
+	vmi := c.resolveControllerVmiRef(pdb.Namespace, controllerRef)
+	if vmi == nil {
+		return
+	}
+	vmiKey, err := controller.KeyFunc(vmi)
+	if err != nil {
+		return
+	}
+	key, err := controller.KeyFunc(pdb)
+	if err != nil {
+		return
+	}
+	c.pdbExpectations.DeletionObserved(vmiKey, key)
+	if err := c.syncPDB(vmi); err != nil {
+		log.Log.V(2).Reason(err).Errorf("could not sync PDB for vmi %v on deleted PDB %v", vmi.Name, pdb.Name)
+	}
+}
+
+func (c *MigrationController) updatePodDisruptionBudget(old, cur interface{}) {
+	curPodDisruptionBudget := cur.(*v1beta1.PodDisruptionBudget)
+	oldPodDisruptionBudget := old.(*v1beta1.PodDisruptionBudget)
+	if curPodDisruptionBudget.ResourceVersion == oldPodDisruptionBudget.ResourceVersion {
+		// Periodic resync will send update events for all known pdbs.
+		// Two different versions of the same pdb will always have different RVs.
+		return
+	}
+
+	if curPodDisruptionBudget.DeletionTimestamp != nil {
+		labelChanged := !reflect.DeepEqual(curPodDisruptionBudget.Labels, oldPodDisruptionBudget.Labels)
+		// having a pdb marked for deletion is enough to count as a deletion expectation
+		c.deletePodDisruptionBudget(curPodDisruptionBudget)
+		if labelChanged {
+			// we don't need to check the oldPodDisruptionBudget.DeletionTimestamp because DeletionTimestamp cannot be unset.
+			c.deletePodDisruptionBudget(oldPodDisruptionBudget)
+		}
+		return
+	}
+
+	curControllerRef := v1.GetControllerOf(curPodDisruptionBudget)
+	oldControllerRef := v1.GetControllerOf(oldPodDisruptionBudget)
+	controllerRefChanged := !reflect.DeepEqual(curControllerRef, oldControllerRef)
+	if controllerRefChanged {
+		// The ControllerRef was changed. Sync the old controller, if any.
+		if vmi := c.resolveControllerVmiRef(oldPodDisruptionBudget.Namespace, oldControllerRef); vmi != nil {
+			if err := c.syncPDB(vmi); err != nil {
```

Comment:
we shouldn't be performing any real work in the pod informer handlers.

The reason for this is, if the syncPDB() funciton returns an error, it won't be processed again. In these informer event handler functions, we need to be queueing a key onto the work queue. With the work-queue, if something like syncPDB fails, then the key is re-enqueued and tried again later.